src/pages/workspaces/workspace/Data.test.ts

```diff
@@ -28,31 +26,49 @@ jest.mock('src/libs/ajax', (): AjaxExports => {
   };
 });
 
+// When Data.js is broken apart and the WorkspaceData component is converted to TypeScript,
+// this type belongs there.
+interface WorkspaceDataProps {
+  namespace: string;
+  name: string;
+  workspace: WorkspaceWrapper;
+  refreshWorkspace: () => void;
+  storageDetails: StorageDetails;
+}
+type AjaxContract = ReturnType<typeof Ajax>;
+
 describe('WorkspaceData', () => {
-  // When Data.js is broken apart and the WorkspaceData component is converted to TypeScript,
-  // this type belongs there.
-  interface WorkspaceDataProps {
+  type SetupOptions = {
     namespace: string;
     name: string;
     workspace: WorkspaceWrapper;
     refreshWorkspace: () => void;
     storageDetails: StorageDetails;
-  }
+    status: LeoAppStatus;
+  };
+  type SetupResult = {
+    workspaceDataProps: WorkspaceDataProps;
+    mockGetSchema: jest.Mock;
+  };
 
-  // SIFERS setup, see: https://medium.com/@kolodny/testing-with-sifers-c9d6bb5b362
-  function setup({
-    namespace = 'test-namespace',
-    name = 'test-name',
-    workspace,
```

Comment:
I deliberately left `workspace` unset in this old style, so that the client tests would _have_ to provide a `workspace` value.  However, by introducing the `SetupOptions` type, I couldn't figure out how to leave `workspace` unset in the invocation of `setup`, which meant adding `...defaultSetupOptions` and sprinkling it throughout.

Is there an idiomatic way for me to preserve the old invocation style, where the individual tests _had_ to provide a workspace, but didn't have to provide an entire `SetupOptions` class?

```setup({ workspace: defaultAzureWorkspace, status: 'RUNNING' })```

instead of:
```setup({ ...defaultSetupOptions, workspace: defaultAzureWorkspace, status: 'RUNNING' })```

Reply:
You could mark the property as optional in the Options type.
https://www.typescriptlang.org/docs/handbook/2/objects.html#optional-properties

Or possibly have the setup function take an argument that's a `Partial<WorkspaceDataProps>` and merge it with the default values.
https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype

Reply:
I dug up an old [resource](https://bobbyhadz.com/blog/typescript-object-parameter-default-value#:~:text=To%20set%20default%20value%20for,all%20its%20properties%20to%20optional) to figure out what I wanted to accomplish here, and pushed up this [attempt](https://github.com/DataBiosphere/terra-ui/pull/4169/commits/8c836ed2349a89f0917bdd7efbdda98732373715) to smooth over this rough edge.